### PR TITLE
Update AbstractArchiveTask to expose Provider API

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Before this change, there was no easy way to rely upon the output
of the AbstractArchiveTask without using dependsOn.

Related: https://github.com/gradle/gradle-native/issues/893

Signed-off-by: Jonathan Leitschuh <Jonathan.Leitschuh@plexxi.com>

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java [hunk 1]
```diff
@@ -170,12 +170,7 @@ public void execute(ScalaCompile scalaCompile) {
                 // TODO: Replace this with Providers in Jar/AbstractArchiveTask
```
Comment (JLLeitschuh) [inline subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java]: Should this TODO be removed now?

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 2]
```diff
@@ -87,41 +98,61 @@ private String maybe(String prefix, String value) {
...
+    }
+
+    /**
+     * The {@link RegularFile} where the archive is constructed.
+     * The path is simply the {@code destinationDir} plus the {@code archiveName}.
+     *
+     * @implNote This returns a provider of {@link RegularFile} instead of {@link RegularFileProperty} in order to
+     * prevent users calling {@link org.gradle.api.provider.Property#set} and causing a plugin or users using
+     * {@link AbstractArchiveTask#getArchivePath()} to break or have strange behaviour.
+     * An example can be found <a href="https://github.com/gradle/gradle-native/issues/893#issuecomment-430776251">here</a>.
```
Comment (JLLeitschuh) [inline subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: I didn't really want this to show up in the generated docs, so I left it using the `implNote` tag, hope that's okay.

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 3]
```diff
@@ -144,38 +176,50 @@ public void setAppendix(String appendix) {
...
-    public void setExtension(String extension) {
-        this.extension = extension;
+
+    /**
+     * Returns the extension part of the archive name.
+     */
+    @Internal
+    public Property<String> getExt() {
+        return extension;
     }
```
Comment (JLLeitschuh) [inline subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: If anyone has a better name for this, let me know.

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 4]
```diff
@@ -28,18 +32,29 @@
...
+import java.util.concurrent.Callable;
 
 /**
  * {@code AbstractArchiveTask} is the base class for all archive tasks.
  */
 public abstract class AbstractArchiveTask extends AbstractCopyTask {
-    private File destinationDir;
+    private DirectoryProperty destinationDirectory = getProject().getObjects().directoryProperty();
+    private final RegularFileProperty archiveFile = getProject().getObjects().fileProperty();
+    {
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: I could if you really wanted me to.
It just seemed clearer to have this logic as close to where the field was declared as possible.

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 5]
```diff
@@ -28,18 +32,29 @@
...
+import javax.annotation.Nullable;
 import java.io.File;
+import java.util.concurrent.Callable;
 
 /**
  * {@code AbstractArchiveTask} is the base class for all archive tasks.
  */
 public abstract class AbstractArchiveTask extends AbstractCopyTask {
-    private File destinationDir;
+    private DirectoryProperty destinationDirectory = getProject().getObjects().directoryProperty();
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: Agreed!

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 6]
```diff
@@ -28,18 +32,29 @@
...
+                return getArchiveName();
+            }
+        })));
+    }
     private String customName;
     private String baseName;
     private String appendix;
     private String version;
-    private String extension;
+    private Property<String> extension = getProject().getObjects().property(String.class);
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: Agreed!

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 7]
```diff
@@ -87,41 +102,61 @@ private String maybe(String prefix, String value) {
...
     public void setDestinationDir(File destinationDir) {
-        this.destinationDir = destinationDir;
+        destinationDirectory.set(destinationDir);
+    }
+
+    /**
+     * The directory where the archive will be placed.
+     */
+    @Internal("Represented by the archiveFile")
+    public DirectoryProperty getDestinationDirectory() {
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: Suser. What is `subprojects/docs/src/docs/dsl/org.gradle.api.tasks.bundling.AbstractArchiveTask.xml`?

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java (hunk 3 above)
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: `getExtension` is already taken, I'll use `getArchiveExtension`.

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java [hunk 8]
```diff
@@ -87,41 +102,61 @@ private String maybe(String prefix, String value) {
      *
      * @return a File object with the path to the archive
      */
-    @OutputFile
+    @Internal("Represented as a part of the archiveFile")
     public File getArchivePath() {
-        File destinationDir = getDestinationDir();
-        //noinspection ConstantConditions
-        if (destinationDir == null) {
-            throw new InvalidUserDataException("The destinationDir property must be set. Please apply the base plugin or set it explicitly.");
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: There's still an exception thrown, it's a different exception now though and it happens at a later phase.

File: subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java]: Done, I think.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/BasePlugin.java [hunk 9]
```diff
@@ -87,35 +86,39 @@ public void apply(final Project project) {
...
+                            }
+                        };
+                    } else {
+                        destinationDir = new Callable<String>() {
+                            public String call() {
+                                return pluginConvention.getDistsDirName();
+                            }
+                        };
+                    }
+                    task.getDestinationDirectory().set(project.getLayout().getBuildDirectory().dir(project.provider(destinationDir)));
```
Comment (JLLeitschuh) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/BasePlugin.java]: Thank you @big-guy for working on this with me!

**Comment by JLLeitschuh** [comment]
Supposedly I introduced some binary incompatibility?? @big-guy can you tell me what that means so I can try and fix it?

**Comment by big-guy** [comment]
Thanks @JLLeitschuh, I merged this locally, fixed a couple of things and pushed it to master.  This'll be in 5.1

**Comment by JLLeitschuh** [comment]
Thanks!

**Pull request closed without merging**